prov/gni/src/gnix_vc.c

```diff
@@ -69,6 +70,367 @@ static int __gnix_vc_conn_ack_comp_fn(void *data);
+				GNIX_WARN(FI_LOG_EP_CTRL,
+					  "_gnix_table_reverse_lookup returned %s\n",
+					  fi_strerror(-ret));
+				return ret;
+			}
+		} else {
+			*table_dest_addr = *((fi_addr_t *) dest_addr);
+
+                        GNIX_INFO(FI_LOG_EP_CTRL, "fi_addr_t: %llx\n",
+                                  *table_dest_addr);
+                }
+
+		ret = _gnix_table_lookup(av, *table_dest_addr, av_entry);
+		assert(*av_entry);
+
+		if (ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_CTRL,
+				  "_gnix_table_lookup returned %s\n",
+				  fi_strerror(-ret));
+			return ret;
+		}
+		/* fastlock_acquire(&ep->vc_lock); */
+                ret = _gnix_vec_at(ep->vc_table, (void **)vc_ptr,
+                                   *table_dest_addr);
+                /* fastlock_release(&ep->vc_lock); */
+                GNIX_DEBUG(FI_LOG_EP_CTRL, "Found vc %p at vector index %llu\n",
+                           *vc_ptr, *table_dest_addr);
+
+                /* Ignore FI_ECANCELED as it's returned when looking up a entry
+                 * that is empty */
+                if (ret != -FI_ECANCELED && ret != FI_SUCCESS) {
+			GNIX_WARN(FI_LOG_EP_CTRL, "_gnix_vec_at returned %s\n",
+				  fi_strerror(-ret));
+                }
+
+		return ret;
+	} else if (av_type == FI_AV_MAP) {
+		if (key_ptr == NULL) {
+			/* Check to see if the user passed in a gnix address */
+			if (is_dest_addr_gnix_addr) {
+				*ht_dest_addr = *((struct gnix_address *)
+						  dest_addr);
+
+                                GNIX_INFO(FI_LOG_EP_CTRL, "gnix_addr: 0x%llx\n",
+					  *ht_dest_addr);
+
+				key_ptr = malloc(sizeof(gnix_ht_key_t));
+				assert(key_ptr);
+
+                                memcpy(key_ptr, ht_dest_addr,
+                                       sizeof(gnix_ht_key_t));
+                                *av_entry =
+                                    _gnix_ht_lookup(av->map_ht, *key_ptr);
+
+                                assert(*av_entry);
+			} else {
+				ret = _gnix_map_lookup(av,
+						       *((fi_addr_t *)dest_addr),
+						       av_entry);
+                                assert(*av_entry);
+                                GNIX_DEBUG(FI_LOG_EP_CTRL,
+					   "av_entry at %p in __gnix_vc_lookup_vc",
+					   *av_entry);
+
+                                if (ret != FI_SUCCESS) {
+					GNIX_WARN(FI_LOG_EP_CTRL,
+						  "_gnix_map_lookup returned %s\n",
+						  fi_strerror(-ret));
+					return ret;
+				}
+				memcpy(ht_dest_addr, &((av_entry[0])->gnix_addr),
+				       sizeof(struct gnix_address));
+
+				key_ptr = malloc(sizeof(gnix_ht_key_t));
+				assert(key_ptr);
+
+                                memcpy(key_ptr, ht_dest_addr,
+                                       sizeof(gnix_ht_key_t));
+
+                                GNIX_INFO(FI_LOG_EP_CTRL,
+					  "fi_addr_t: 0x%llx gnix_addr: 0x%llx\n",
+					  *((fi_addr_t *) dest_addr),
+					  (av_entry[0])->gnix_addr);
+			}
+
+		}
+		/* fastlock_acquire(&ep->vc_lock); */
```

Comment:
I’m not sure, this is something I would like to discuss.  Currently the lock is taken before calling __gnix_vc_lookup_vc, but, I think, the lock only needs to be taken within __gnix_vc_lookup_vc when reading from the hashtable or vector.  Perhaps this code would perform better if we uncommented the fastlock calls within __gnix_vc_lookup_vc and moved the existing fastlock_acquire call after the __gnix_vc_lookup_vc returns.  Any thoughts?
> On Mar 28, 2016, at 9:00 AM, Howard Pritchard <notifications@github.com> wrote:
> 
> In prov/gni/src/gnix_vc.c <https://github.com/ofi-cray/libfabric-cray/pull/736#discussion_r57580621>:
> 
> > +				       sizeof(struct gnix_address));
> > +
> > +				key_ptr = malloc(sizeof(gnix_ht_key_t));
> > +				assert(key_ptr);
> > +
> > +                                memcpy(key_ptr, ht_dest_addr,
> > +                                       sizeof(gnix_ht_key_t));
> > +
> > +                                GNIX_INFO(FI_LOG_EP_CTRL,
> > +					  "fi_addr_t: 0x%llx gnix_addr: 0x%llx\n",
> > +					  *((fi_addr_t *) dest_addr),
> > +					  (av_entry[0])->gnix_addr);
> > +			}
> > +
> > +		}
> > +		/* fastlock_acquire(&ep->vc_lock); */
> do we need the lock to be taken here?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub <https://github.com/ofi-cray/libfabric-cray/pull/736/files/408e2a2ddff4bc9751170dc9ac6f6a66a970679b#r57580621>